ios/engine/KMEI/KeymanEngine/Classes/KeyPreviewView.swift

```diff
@@ -0,0 +1,193 @@
+                   tangent2End: CGPoint(x: viewLeft, y: midY2), radius: r1)
+    context.addArc(tangent1End: CGPoint(x: viewLeft, y: midY2),
+                   tangent2End: CGPoint(x: keyLeft, y: midY1), radius: r2)
+    context.addArc(tangent1End: CGPoint(x: keyLeft, y: midY1),
+                   tangent2End: CGPoint(x: keyLeft, y: viewBottom), radius: r2)
+    context.addArc(tangent1End: CGPoint(x: keyLeft, y: viewBottom),
+                   tangent2End: CGPoint(x: keyRight, y: viewBottom), radius: r0)
+    context.addArc(tangent1End: CGPoint(x: keyRight, y: viewBottom),
+                   tangent2End: CGPoint(x: keyRight, y: midY1), radius: r0)
+    context.addArc(tangent1End: CGPoint(x: keyRight, y: midY1),
+                   tangent2End: CGPoint(x: viewRight, y: midY2), radius: r2)
+    context.addArc(tangent1End: CGPoint(x: viewRight, y: midY2),
+                   tangent2End: CGPoint(x: viewRight, y: viewTop), radius: r2)
+    context.addArc(tangent1End: CGPoint(x: viewRight, y: viewTop),
+                   tangent2End: CGPoint(x: viewMid, y: viewTop), radius: r1)
+    context.closePath()
+    context.drawPath(using: .stroke)
+
+    context.beginPath()
+    context.move(to: CGPoint(x: viewMid, y: viewTop))
+    context.addArc(tangent1End: CGPoint(x: viewLeft, y: viewTop),
+                   tangent2End: CGPoint(x: viewLeft, y: midY2), radius: r1)
+    context.addArc(tangent1End: CGPoint(x: viewLeft, y: midY2),
+                   tangent2End: CGPoint(x: keyLeft, y: midY1), radius: r2)
+    context.addArc(tangent1End: CGPoint(x: keyLeft, y: midY1),
+                   tangent2End: CGPoint(x: keyLeft, y: viewBottom), radius: r2)
+    context.addArc(tangent1End: CGPoint(x: keyLeft, y: viewBottom),
+                   tangent2End: CGPoint(x: keyRight, y: viewBottom), radius: r0)
+    context.addArc(tangent1End: CGPoint(x: keyRight, y: viewBottom),
+                   tangent2End: CGPoint(x: keyRight, y: midY1), radius: r0)
+    context.addArc(tangent1End: CGPoint(x: keyRight, y: midY1),
+                   tangent2End: CGPoint(x: viewRight, y: midY2), radius: r2)
+    context.addArc(tangent1End: CGPoint(x: viewRight, y: midY2),
+                   tangent2End: CGPoint(x: viewRight, y: viewTop), radius: r2)
+    context.addArc(tangent1End: CGPoint(x: viewRight, y: viewTop),
+                   tangent2End: CGPoint(x: viewMid, y: viewTop), radius: r1)
+    context.closePath()
+    context.clip()
+
+    let colorSpace = CGColorSpaceCreateDeviceRGB()
+    let gradientColors = [bgColor.cgColor, bgColor2.cgColor] as CFArray
+    let gradientLocations: [CGFloat] = [0, 1]
+    let gradient = CGGradient(colorsSpace: colorSpace, colors: gradientColors, locations: gradientLocations)!
+
+    let startPoint = CGPoint(x: rect.midX, y: rect.minY)
+    let endPoint = CGPoint(x: rect.midX, y: rect.maxY)
+    context.drawLinearGradient(gradient, start: startPoint, end: endPoint, options: [])
+    UIGraphicsPopContext()
```

Comment:
Swift auto-GCs the gradient and color space objects for us now, right?

Reply:
Yes. It's an error if we try to manually release.